### PR TITLE
Add trusted types in client side Angular 

### PR DESCRIPTION
**Pull request by Siegrift**
## Trusted Types
[Trusted Types](https://github.com/WICG/trusted-types) ([spec](https://wicg.github.io/trusted-types/dist/spec/), [introductory article](https://developers.google.com/web/updates/2019/02/trusted-types)) is a new experimental DOM API implemented within the WICG , with a working [Chrome implementation](https://www.chromestatus.com/feature/5650088592408576).

The API creates a few new objects available on the global object in the browser, like most other web APIs ([impl in TS](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/trusted-types/index.d.ts) and in [Closure compiler](https://github.com/WICG/trusted-types/blob/master/externs/externs.js)).

Under certain conditions, controlled by a HTTP header (analogous to Content-Security-Policy behavior), the API can enable the enforcement - then it changes the signature of several DOM API functions and property setters, such that they accept specific object types, and reject strings. Colloquially, DOM API becomes strongly typed.

For example, with Trusted Types Element.innerHTML property setter accepts a TrustedHTML object.

Trusted Type objects stringify to their inner value. This API shape is a deliberate choice that enables existing web applications and libraries to gradually migrate from strings to Trusted Types without breaking functionality. In our example, it makes it possible to write the following:

```javascript
const policy = TrustedTypes.createPolicy('foo', { 
  createHTML: (s) => { /* some validation*/; return s} 
});

const trustedHTML = policy.createHTML('bar');
anElement.innerHTML = trustedHTML

anElement.innerHTML === 'bar'
```

The above code works regardless if the Trusted Types enforcement is enabled or not.

Reading from the DOM is unaffected, so Element.innerHTML getter returns a string. That's for practical reasons -- web applications read from DOM more often than they write to it, and only writing exposes the application to DOM XSS risks. Typing only the setters allows us to secure web applications with minimal code changes.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit _(Although, it would be nice to squash them)_
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**This is only a draft PR for now, there are basically zero tests written and the PR is only tested locally on my sample application. In next few days I will add more tests, and I can squash the commits if you want want.**

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently, there is no support for Trusted Types in Angular.

## What is the new behavior?

If Trusted Types (TT) are available in the browser, they will be used out of the box. Users are able to configure policy name created by Angular to wrap sanitized values. There is a way in Angular to Angular skips sanitization not only for [values created using bypass functions](https://angular.io/api/platform-browser/DomSanitizer#security-risk), but also for trusted values created by TT policies. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No - Feature is only enabled when Trusted Types are available in the browser, but the API doesn't enforce anything unless users set the enforcing mode through CSP header.

## Other information
1) TT API is only a browser API and if application uses [universal rendering](https://angular.io/guide/universal) it may introduce vulnerabilities that could be avoided by rendering on client side (with TT enforced). We could possibly emulate TT on the server, but this requires much more work and it's possibly best to add TT to [domino](https://github.com/fgnass/domino). This PR is **only** focusing on client side and there is no support for TT on server side yet.
2) Applications which use [internationalization](https://angular.io/guide/i18n) will break, because of the hacky functions accessing DOM APIs inside [inert_body.ts](https://github.com/angular/angular/blob/master/packages/core/src/sanitization/inert_body.ts), which are called by internationalization module [here](https://github.com/angular/angular/blob/master/packages/core/src/render3/i18n.ts#L1165).

### CODE REVIEW

File: packages/core/src/security/trusted_types_policy.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  constructor(@Optional() @Inject(DOM_SANITIZATION_POLICY_NAME) private policyName: string|
+              undefined) {
+    super();
+    this.tt = window.trustedTypes || (window as any).TrustedTypes;
+    if (this.tt && this.tt.createPolicy) {
+      this.policy = this.tt.createPolicy(this.policyName || 'angular-sanitization', {
+        createScriptURL: (s: string) => s,
+        createScript: (s: string) => s,
+        createHTML: (s: string) => s
+      });
```
Comment (mhevery) [inline packages/core/src/security/trusted_types_policy.ts]: All of these functions are identity. I think I would expect to see the actual sanitization code inside there

**Review by mhevery — CHANGES_REQUESTED** [review]
(I am sorry that I was not able to review this earlier and provide more timely feedback)

Primary concern for this PR is that it is not compatible with the spirit of Ivy (render3). In Ivy we wanted to make sure that the injection system is optional, and as much of the system is tree-shakable as possible. This implementation relies on Injection as well as it is not tree-shakable because bringing any part of sanitization, will bring the whole sanitization into the codebase. See how [ivy retrieves sanitizer](https://github.com/angular/angular/blob/3a53e2c960220182ff6dea13708d4e35d0a9bc76/packages/core/src/sanitization/sanitization.ts#L90-L99). Notice that it is optional and tree shakable. (If the generated instructions don't bring in `ɵɵsanitizeUrl` then the sanitization system is not shipped with the application)

Secondary concern is that [`createPolicy`](https://github.com/angular/angular/pull/32353/files#r331650067) uses identity functions. To follow the spirit of the TrustedTypes, I would expect that the actual sanitization code would be invoked from there. To me it feels like TrustedTypes were added to Angular sanitization system to make TrustedTypes happy rather than to follow the spirit of TrustedTypes. 

It is not obvious to me what is the best way to integrate TrustedTypes in a way which follows the spirit of TrustedTypes as well as Ivy and tree-shakability. I think we should have a design doc where we discuss these trade-offs and constraints before we decided on how best to refactor it.

**Comment by wibimaster** [comment]
Hi ! 

Is there any news, after 9 months ?

I opened a question no StackOverflow before seeing this PR :
https://stackoverflow.com/questions/62810553/how-to-fix-trustedhtml-assignment-error-with-angular-innerhtml

Maybe you, who understand the trustedTypes model, can help us to find why we cannot pass through the Chrome errors ?

Thanks !

**Comment by koto** [comment]
The issue is that without this patch, Angular will stringify the TrustedHTML instance, causing the Trusted Type violation. We're currently working on a new integration, as the rendering engine (Ivy) obsoleted this patch. 

Currently the only way to run an Angular application under Trusted Types is to use a [default policy](https://w3c.github.io/webappsec-trusted-types/dist/spec/#default-policy-hdr), or hard-fork the framework code, but that's hardly a recommended solution :(

**Comment by IgorMinar** [comment]
closing in favor of #39222 most of which had already been merged. thank you

**Pull request closed without merging**